TextInputKit/TextInputKit/Code/SpecializedTextInput/StringUtils.swift

```diff
@@ -0,0 +1,119 @@
+//
+//  StringUtils.swift
+//  TextInputKit
+//
+//  Created by Artem Starosvetskiy on 10/12/2016.
+//  Copyright © 2016 Artem Starosvetskiy. All rights reserved.
+//
+
+import Foundation
+
+struct StringUtils {
+
+    private init() {}
+
+}
+
+extension StringUtils {
+
+    static func stringView(
+        from stringView: String.UnicodeScalarView,
+        preservingCharacters: CharacterSet,
+        ignoringCharacters: CharacterSet
+        ) -> String.UnicodeScalarView? {
+
+        do {
+            var resultStringView = "".unicodeScalars
+            try scan(stringView,
+                     preservingCharacters: preservingCharacters,
+                     ignoringCharacters: ignoringCharacters,
+                     appending: &resultStringView)
+
+            return resultStringView
+        }
+        catch {
+            return nil
+        }
+    }
+
+    static func stringViewAndRange(
+        from originalStringView: String.UnicodeScalarView,
+        range originalRange: Range<String.UnicodeScalarIndex>,
+        preservingCharacters: CharacterSet,
+        ignoringCharacters: CharacterSet
+        ) -> (String.UnicodeScalarView, Range<String.UnicodeScalarIndex>)? {
+
+        precondition((originalRange.lowerBound == originalStringView.endIndex) || originalStringView.indices.contains(originalRange.lowerBound))
+        precondition((originalRange.upperBound == originalStringView.endIndex) || originalStringView.indices.contains(originalRange.upperBound))
+
+        var resultStringView = "".unicodeScalars
+
+        do {
+            if originalRange.lowerBound != originalStringView.startIndex {
+                let leftSubstringView = (originalRange.lowerBound == originalStringView.endIndex)
+                    ? originalStringView
+                    : originalStringView.prefix(upTo: originalRange.lowerBound)
+                try scan(
+                    leftSubstringView,
+                    preservingCharacters: preservingCharacters,
+                    ignoringCharacters: ignoringCharacters,
+                    appending: &resultStringView)
+            }
+
+            let resultRangeLowerBound = resultStringView.endIndex
+
+            if !originalRange.isEmpty {
+                let middleSubstringView = (originalRange.lowerBound == originalStringView.startIndex) && (originalRange.upperBound == originalStringView.endIndex)
```

Comment:
First of all, it can be written in more "friendly" way:
Separate var for condition.

Also at the moment swift compiler feels bad seeing things like that :)